libethashseal/EthashClient.cpp

```diff
@@ -59,6 +59,11 @@ EthashClient::EthashClient(
 	asEthashClient(*this);
 }
 
+EthashClient::~EthashClient()
+{
+	terminate();
```

Comment:
So it will be called twice - once in `~EthashClient`, then in `~Client` ?

Reply:
Yes, which is not a problem.  I'll comment about that.

Reply:
Done.

Reply:
I think not calling `terminate` here shouldn't cause any problems, because no virtual functions are overriden in `EthashClient`
But if it fixes some TSAN warnings, then fine.

Also note that there's `ClientTest` class without this call, too.

Reply:
Without this change I see a warning:
```
WARNING: ThreadSanitizer: data race on vptr (ctor/dtor vs virtual call) (pid=21629)
  Write of size 8 at 0x7ba400000550 by main thread:
    #0 ~Client /home/yh/src/cpp-ethereum2/libethereum/Client.cpp:82 (eth+0x697c0d)
    #1 ~EthashClient /home/yh/src/cpp-ethereum2/libethashseal/EthashClient.h:37 (discriminator 2) (eth+0x7b4021)
    #2 ~EthashClient /home/yh/src/cpp-ethereum2/libethashseal/EthashClient.h:37 (eth+0x7b4049)
    #3 std::default_delete<dev::eth::Client>::operator()(dev::eth::Client*) const /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/bits/unique_ptr.h:76 (discriminator 1) (eth+0x83039f)
    #4 std::unique_ptr<dev::eth::Client, std::default_delete<dev::eth::Client> >::reset(dev::eth::Client*) /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/bits/unique_ptr.h:347 (eth+0x82e179)
    #5 ~WebThreeDirect /home/yh/src/cpp-ethereum2/libwebthree/WebThree.cpp:91 (eth+0x82d596)
    #6 main /home/yh/src/cpp-ethereum2/eth/main.cpp:1262 (discriminator 15) (eth+0x514a19)

  Previous read of size 8 at 0x7ba400000550 by thread T13:
    #0 dev::Worker::workLoop() /home/yh/src/cpp-ethereum2/libdevcore/Worker.cpp:140 (eth+0x852597)
    #1 operator() /home/yh/src/cpp-ethereum2/libdevcore/Worker.cpp:62 (eth+0x852afd)
    #2 void std::_Bind_simple<dev::Worker::startWorking()::$_0 ()>::_M_invoke<>(std::_Index_tuple<>) /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/functional:1390 (discriminator 2) (eth+0x852989)
    #3 std::_Bind_simple<dev::Worker::startWorking()::$_0 ()>::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/functional:1380 (eth+0x852939)
    #4 std::thread::_State_impl<std::_Bind_simple<dev::Worker::startWorking()::$_0 ()> >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/thread:197 (eth+0x8527fd)
    #5 std::error_code::default_error_condition() const ??:? (libstdc++.so.6+0xbb83e)
```

Myaybe the destructor was forwarded or something.

Reply:
Ok I see, vptr is probaly being modified somewhere between `~EthashClient` & `~Client` body and worker thread can access it at this moment, that's the problem. 
Let's leave `terminate()` here. And add it to `ClientTest`

Reply:
Done.